Sources/Target/Linux/Process.cpp

```diff
@@ -215,19 +215,54 @@ ErrorCode Process::wait() {
 
     switch (_currentThread->_stopInfo.event) {
     case StopInfo::kEventNone:
-      // If the thread is stopped for no reason, it means the debugger (ds2)
-      // sent a SIGSTOP to it while it was already stopped for another reason,
-      // the SIGSTOP was queued, and now we're getting this notification. We
-      // have to ignore this and just let the thread continue.
-      // If `stepping` is true, it means the thread was actually being
-      // single-stepped before stopping, so instead of doing a `resume()`, we
-      // have to do a new `step()`.
-      if (stepping) {
+      // If the _stopInfo.event is kEventNone it means the thread is either
+      // 1. stopped because a sibling thread hit a breakpoint and ds2
+      //    sent a SIGSTOP to pause this thread while the breakpoint is
+      //    evaluated. In this case we just call continue.
+      // 2. stopped because this thread called clone(3). In this case
+      //    we have two subcases:
+      //    a. The thread was running and hit the clone, stopped
+      //       and we just continue it and continue the child as well.
+      //       The loop will continue and the next waitpid will handle the
+      //       freshly cloned thread. In this case we we call continue as well
+      //    b. We were single stepping (c.f. the `stepping` var) and
+      //       were stopped by a SIGTRAP for the step while simultaneously
+      //       being stopped via the SIGTRAP for the clone.
+      //       In this case we can run into a race condition with the
+      //       waitpid(3) call and thus have to handle the cloned thread
+      //       and the currentThread at once. So we call step and then
+      //       beforeResume the cloned thread.
+      //
+      // We also have another theoretically possible case where we are not
+      // stopped by a clone but still stepping while encountering a kEventNone.
+      // This should not happen and we just assert that it does not.
+      //
+      if (_currentThread->_stopInfo.reason == StopInfo::kReasonClone &&
```

Comment:
You can probably re-order this if/else if/else block to avoid having more than one `goto`, and check early for the fatal error you have at line 259.